src/ray/common/cgroup/cgroup_setup.cc

```diff
@@ -73,14 +80,55 @@ bool MoveProcsInSystemCgroup() {
   return out_file.good();
 }
 
+// This function is the reverse operation for [MoveProcsInSystemCgroup], which moves all
+// processes from raylet system cgroup into root cgroup.
+//
+// Return whether the PIDs move successfully.
+bool MoveProcsFromSystemCgroup() {
+  std::ifstream in_file(cgroup_v2_system_folder.data());
+  std::ofstream out_file(kRtootCgroupProcs.data());
+  int pid = 0;
+  while (in_file >> pid) {
+    out_file << pid << std::endl;
+  }
+  return out_file.good();
+}
+
 // Return whether cgroup control writes successfully.
 bool EnableCgroupSubtreeControl(const char *subtree_control_path) {
   std::ofstream out_file(subtree_control_path);
-  // Able to add new PIDs and memory constraint to the system cgroup.
-  out_file << "+memory +pids";
+  // Able to add memory constraint to the system cgroup.
+  out_file << "+memory";
   return out_file.good();
 }
 
+// Kill all processes inside of the cgroup.
+// If any operation fails, the error is logged and continue.
+void KillProcsUnderCgroupFolder(const std::string &folder) {
+  const std::string cgroup_v2_app_procs = ray::JoinPaths(folder, "cgroup.procs");
+  std::vector<int> dangling_pids;
+  std::ifstream in_file(cgroup_v2_app_procs.data());
+  int pid = 0;
+  while (in_file >> pid) {
+    dangling_pids.emplace_back(pid);
+  }
+
+  for (int pid : dangling_pids) {
+    if (kill(pid, SIGKILL) == -1) {
```

Comment:
should we SIGTERM with a grace period first?

if this always immediately SIGKILLs, please document it in the header comment (not expected behavior)

Reply:
The function is expected to invoke only at raylet destruction, at which point we expect all worker processes to complete and exit. But it's a good call to clearly document the behavior.

Reply:
Some system uses SIGTERM because they have sigterm handler and a drain mechanism, which rejects all new incoming requests and gracefully shutdown, but I don't think it's something applicable to ray worker process.